### PR TITLE
Aha/firewall

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/web/http/Firewall.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+package sirius.web.http;
+
+/**
+ * Permits application to provide a firewall which can perform IP filtering and rate limiting.
+ */
+public interface Firewall {
+
+    /**
+     * Determines if the remote IP of the given requrest is balcklisted
```
Comment (sabieber) [inline src/main/java/sirius/web/http/Firewall.java]: requrest -> request
balcklisted -> blacklisted

File: src/main/java/sirius/web/controller/ControllerDispatcher.java [hunk 2]
```diff
@@ -71,79 +72,92 @@ public int getPriority() {
     }
 
     @Override
-    public boolean preDispatch(WebContext ctx) throws Exception {
-        return route(ctx, true);
-    }
+    @SuppressWarnings("squid:S1698")
+    @Explain("We acutally can use object identity here as this is a marker object.")
```
Comment (sabieber) [inline src/main/java/sirius/web/controller/ControllerDispatcher.java]: acutally -> actually

File: src/main/java/sirius/web/controller/ControllerDispatcher.java [hunk 3]
```diff
@@ -71,79 +72,92 @@ public int getPriority() {
...
+    public Callback<WebContext> preparePreDispatch(WebContext ctx) {
+        String uri = determineEffectiveURI(ctx);
+        for (final Route route : getRoutes()) {
+            final List<Object> params = shouldExecute(ctx, uri, route, true);
+            if (params != Route.NO_MATCH) {
+                InputStreamHandler ish = new InputStreamHandler();
```
Comment (sabieber) [inline src/main/java/sirius/web/controller/ControllerDispatcher.java]: ish -> handler or inputStreamHandler?

File: src/main/java/sirius/web/controller/ControllerDispatcher.java [hunk 4]
```diff
@@ -71,79 +72,92 @@ public int getPriority() {
...
 
     @SuppressWarnings("squid:S1698")
-    @Explain("Strings have been intered for performance reasons.")
-    private boolean tryExecuteRoute(WebContext ctx, boolean preDispatch, String uri, Route route) {
-        try {
-            final List<Object> params = route.matches(ctx, uri, preDispatch);
-            if (params == Route.NO_MATCH) {
-                // Route did not match...
-                return false;
+    @Explain("We acutally can use object identity here as this is a marker object.")
```
Comment (sabieber) [inline src/main/java/sirius/web/controller/ControllerDispatcher.java]: acutally -> actually

File: src/main/java/sirius/web/controller/ControllerDispatcher.java [hunk 5]
```diff
@@ -71,79 +72,92 @@ public int getPriority() {
...
-                                                 "Request dropped - System overload!"))
-                 .fork(() -> handleFailure(ctx, route, e));
-            return true;
+            handleFailure(ctx, route, e);
         }
     }
 
+    @Override
+    @SuppressWarnings("squid:S1698")
+    @Explain("We acutally can use object identity here as this is a marker object.")
```
Comment (sabieber) [inline src/main/java/sirius/web/controller/ControllerDispatcher.java]: acutally -> actually

File: src/main/java/sirius/web/http/DispatcherPipeline.java [hunk 6]
```diff
@@ -60,7 +68,13 @@ public static DispatcherPipeline create() {
      * @param ctx the request to handle
      */
     public void dispatch(WebContext ctx) {
-        dispatch(ctx, this::dispatch, CallContext.getCurrent().get(TaskContext.class));
+        tasks.executor(EXECUTOR_WEBSERVER)
```
Comment (tbiScireum) [inline src/main/java/sirius/web/http/DispatcherPipeline.java]: unit test ?

File: src/main/java/sirius/web/controller/ControllerDispatcher.java [hunk 7]
```diff
@@ -115,25 +141,29 @@ private boolean tryExecuteRoute(WebContext ctx, boolean preDispatch, String uri,
...
-                                                 "Request dropped - System overload!"))
-                 .fork(() -> handleFailure(ctx, route, e));
-            return true;
+            handleFailure(ctx, route, e);
+        }
+    }
+
+    @Override
+    @SuppressWarnings("squid:S1698")
+    @Explain("We acutally can use object identity here as this is a marker object.")
```
Comment (Tobsucht) [inline src/main/java/sirius/web/controller/ControllerDispatcher.java]: actually

File: src/main/java/sirius/web/http/Firewall.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ * by scireum in Remshalden, Germany
+ *
+ * Copyright by scireum GmbH
+ * http://www.scireum.de - info@scireum.de
+ */
+
+package sirius.web.http;
+
+/**
+ * Permits application to provide a firewall which can perform IP filtering and rate limiting.
```
Comment (Tobsucht) [inline src/main/java/sirius/web/http/Firewall.java]: an application

File: src/main/java/sirius/web/controller/ControllerDispatcher.java [hunk 9]
```diff
@@ -71,42 +79,60 @@ public int getPriority() {
...
+                                     Route route,
+                                     List<Object> params,
+                                     InputStreamHandler inputStreamHandler) {
+        try {
+            if (firewall != null && !route.getMethod().isAnnotationPresent(Unlimited.class)) {
+                if (firewall.handleRateLimiting(ctx,
+                                                Optional.ofNullable(route.getMethod().getAnnotation(Limited.class))
+                                                        .map(Limited::value)
+                                                        .orElse(Limited.HTTP))) {
+                    return;
```
Comment (D3nnisH) [inline src/main/java/sirius/web/controller/ControllerDispatcher.java]: HTTP Status Code 429 "Too many requests" statt einfach ins leere laufen zu lassen?

File: src/main/java/sirius/web/services/ServiceDispatcher.java [hunk 10]
```diff
@@ -108,6 +103,16 @@ private boolean doDispatch(WebContext ctx) {
     private void invokeService(WebContext ctx, ServiceCall call, StructuredService serv) {
         TaskContext.get().setSystem(SYSTEM_SERVICE).setSubSystem(serv.getClass().getSimpleName());
 
+        // Check firewall
+        if (firewall != null && !serv.getClass().isAnnotationPresent(Unlimited.class)) {
+            if (firewall.handleRateLimiting(ctx,
+                                            Optional.ofNullable(serv.getClass().getAnnotation(Limited.class))
+                                                    .map(Limited::value)
+                                                    .orElse(Limited.HTTP))) {
+                return;
```
Comment (D3nnisH) [inline src/main/java/sirius/web/services/ServiceDispatcher.java]: HTTP Status Code 429 "Too many requests" statt einfach ins leere laufen zu lassen?

File: src/main/java/sirius/web/http/DispatcherPipeline.java (hunk 6 above)
Comment (andyHa) [reply src/main/java/sirius/web/http/DispatcherPipeline.java]: done

File: src/main/java/sirius/web/controller/ControllerDispatcher.java (hunk 9 above)
Comment (andyHa) [reply src/main/java/sirius/web/controller/ControllerDispatcher.java]: If handle returns true, we expect the firewall to have the request handled (e.g. with said status code)

File: src/main/java/sirius/web/services/ServiceDispatcher.java (hunk 10 above)
Comment (andyHa) [reply src/main/java/sirius/web/services/ServiceDispatcher.java]: If handle returns true, we expect the firewall to have the request handled (e.g. with said status code)

**Pull request merged**